### PR TITLE
rmfakecloud: build the web ui

### PR DESCRIPTION
**Pull request by i-like-noodles**
###### Description of changes

Previously rmfakecloud was built without the web ui making it show 404 when attempting to use it. Build it similar to how other projects using yarn are built. 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by pacien** [comment]
Thanks for doing this!

Would it be possible to split the build of the web UI into its own
package, whose output would then be included when building the Go part?

This would allow:
- avoiding mixing the two (Node and Go) build environments,
- avoiding rebuilding the web UI when patching the Go sources,
- making the web UI optional more easily.

File: pkgs/servers/rmfakecloud/default.nix [hunk 1]
```diff
@@ -1,21 +1,42 @@
...
+  src = source;
+
   vendorSha256 = "sha256-NwDaPpjkQogXE37RGS3zEALlp3NuXP9RW//vbwM6y0A=";
 
-  postPatch = ''
-    # skip including the JS SPA, which is difficult to build
-    sed -i '/go:/d' ui/assets.go
+  nativeBuildInputs = [ fixup_yarn_lock yarn nodejs ];
+
+  # inspired by https://github.com/NixOS/nixpkgs/blob/358bf0f44fbf320ed64082fceda1ff57aab9ec83/pkgs/servers/peertube/default.nix
```
Comment (SuperSandro2000) [inline pkgs/servers/rmfakecloud/default.nix:28]:
```suggestion
```

File: pkgs/servers/rmfakecloud/default.nix [hunk 2]
```diff
@@ -1,21 +1,42 @@
...
+  preBuild = ''
+    export HOME=$PWD
```
Comment (SuperSandro2000) [inline pkgs/servers/rmfakecloud/default.nix]:
```suggestion
    export HOME=$(mktemp -d)
```

File: pkgs/servers/rmfakecloud/default.nix [hunk 3]
```diff
@@ -1,21 +1,42 @@
...
+    fixup_yarn_lock ~/ui/yarn.lock
+    cd ~/ui
+    yarn config --offline set yarn-offline-mirror ${yarnOfflineCache}
+    yarn install --offline --frozen-lockfile --ignore-engines --ignore-scripts --no-progress
+    patchShebangs ~/ui/node_modules
+    export PATH=~/ui/node_modules/.bin:$PATH
+    ~/ui/node_modules/.bin/react-scripts build
+    mkdir -p $out
+    cd ~
```
Comment (SuperSandro2000) [inline pkgs/servers/rmfakecloud/default.nix]: Can we use relative directories or $HOME instead of tilde?

**Comment by i-like-noodles** [comment]
The web ui is now built with yarn in a separate package and then included. I initially didn't see the suggested changes, but that's in there too now!

**Comment by martinetd** [comment]
> could you make the packages for the server and the web UI share the same source? (To avoid any chance of version mismatch when updating the package in the future).

Which won't be far, 0.0.9 update in the pipes : https://github.com/NixOS/nixpkgs/pull/190166 :)

I'm really scared of node stuff after tons of troubles with cryptpad (currently can't upgrade one of my systems due to it...), so just one question from me: how is one supposed to update the yarn cache/lock thing? Same as package rebuild with a bad hash and fix the hash should just work™?

That aside I don't need the webui at all here (using curl directly when needed), but I guess this makes sense to keep on by default; it's not like it adds extra endpoints... Thanks for this work!

**Comment by i-like-noodles** [comment]
Goddamnit, I messed up the rebase. All I wanted was to pull the version update you added. Git is too complicated.

On the bright side, the package now has the webui as optional and it depends on the version in the main package!

Yes, I just edit the sha256 to be invalid and then rebuild to get the updated ones for the yarnOfflineCache, there's probably a better way but I'm not smart enough clearly.

**Comment by martinetd** [comment]
> Goddamnit, I messed up the rebase. All I wanted was to pull the version update you added. Git is too complicated.

Looks like you did a number on it yes! :)
Let's get that fixed: The easiest is just to shove it away and cherry-pick it back from top of master; keeping the individual commits don't make much sense in your case.

The following commands worked for me after checking out your branch:
```
(create a temporary new branch to make sure we don't lose your work)
$ git branch rm_tmp HEAD
(throw away everything)
$ git reset --hard origin/master
(get your changes back -- the -p lets you select a chunk at a time,
 here we can just take everything so you could skip the -p but it's good practice)
$ git checkout -p rm_tmp pkgs/servers/rmfakecloud
(commit message anew)
$ git commit
(push again to this PR branch)
$ git push -f yourremote yourbranch
(remove temporary old branch)
$ git branch -D rm_tmp
```

In general you want to rebase your branch on top of master, so with your branch checked out `git rebase -i origin/master` will do what you want.
You just somehow did it the other way around, so all the commits that came in master since then got added to your branch, adding lots of people in Cc to this PR. Let's try not to do that too often -- sorry folks, please unsubscribe :(

> On the bright side, the package now has the webui as optional and it depends on the version in the main package!

Thanks!
It looks good, just one question: why did the sed -i become assets.buildGoModule ? changed with the version?

> Yes, I just edit the sha256 to be invalid and then rebuild to get the updated ones for the yarnOfflineCache, there's probably a better way but I'm not smart enough clearly.

That works for me ;)

**Comment by i-like-noodles** [comment]
Evidently I did do it in the wrong order, I pressed sync in the github ui and then pulled that to rebase. Sorry for the mess and sorry for the pings!

Anyway, it looks like it's ok again now, there's only one change, so thanks for the steps needed to undo all those things. It's a good thing you're paying attention too, because the sed line was caused by an accidental autocompletion, it shouldn't be there. It is now how it should be.

**Review by martinetd — APPROVED** [review]
Didn't test resulting binaries but it builds and LGTM on principle. Thanks!

**Pull request merged**